### PR TITLE
Clean up compute() function of RepositoryDelegatorFunction

### PR DESCRIPTION
**Pull request by irengrig**
just a pure refactoring without any new functionality

### CODE REVIEW

File: src/main/java/com/google/devtools/build/lib/rules/repository/RepositoryDelegatorFunction.java [hunk 1]
```diff
@@ -114,7 +112,7 @@ public RepositoryDelegatorFunction(
   private void setupRepositoryRoot(Path repoRoot) throws RepositoryFunctionException {
     try {
       FileSystemUtils.deleteTree(repoRoot);
-      FileSystemUtils.createDirectoryAndParents(repoRoot.getParentDirectory());
+      Preconditions.checkNotNull(repoRoot.getParentDirectory()).createDirectoryAndParents();
```
Comment (lberki) [inline src/main/java/com/google/devtools/build/lib/rules/repository/RepositoryDelegatorFunction.java]: nit: the precondition check is not necessary. If the argument is null, the NullPointerException will contain the exact same information as the exception thrown by the precondition check (this is not the only such place)

File: src/main/java/com/google/devtools/build/lib/rules/repository/RepositoryDelegatorFunction.java [hunk 2]
```diff
@@ -151,91 +144,55 @@ public SkyValue compute(SkyKey skyKey, Environment env)
...
-      // declare a Skyframe dependency on the repository root.
-      RepositoryFunction.getRepositoryDirectory(repoRoot, env);
-      if (env.valuesMissing()) {
-        return null;
+    DigestWriter digestWriter = new DigestWriter(directories, repositoryName, rule,
+        Preconditions.checkNotNull(ruleSpecificData));
+
+    // Local repositories are fetched regardless of the marker file because the operation is
+    // generally fast and they do not depend on non-local data, so it does not make much sense to
+    // try to cache from across server instances.
```
Comment (lberki) [inline src/main/java/com/google/devtools/build/lib/rules/repository/RepositoryDelegatorFunction.java]: nit: try to cache *them* across server instances

File: src/main/java/com/google/devtools/build/lib/rules/repository/RepositoryDelegatorFunction.java [hunk 3]
```diff
@@ -151,91 +144,55 @@ public SkyValue compute(SkyKey skyKey, Environment env)
...
+    boolean fetchLocalRepositoryAlways = isFetch.get() && handler.isLocal(rule);
+    if (!fetchLocalRepositoryAlways) {
+      if (doNotFetchUnconditionally && repoRoot.exists()) {
+        byte[] markerHash = digestWriter.isFileSystemUpToDate(handler, env);
```
Comment (lberki) [inline src/main/java/com/google/devtools/build/lib/rules/repository/RepositoryDelegatorFunction.java]: What happened to the comment in line 192 that tells why we can't depend on `FileValues` that fetching this repository would overwrite? Passing `SkyFunction.Environment` to a method called `isFileSystemUpToDate` is at least suspicious in that respect. Or that comment was wrong for some reason?

(not that HEAD doesn't have this problem, but you seem to have deleted the comment for some reason, which makes me think that something was wrong about it)

File: src/main/java/com/google/devtools/build/lib/rules/repository/RepositoryDelegatorFunction.java [hunk 4]
```diff
@@ -151,91 +144,55 @@ public SkyValue compute(SkyKey skyKey, Environment env)
...
-      RepositoryFunction.getRepositoryDirectory(repoRoot, env);
-      if (env.valuesMissing()) {
-        return null;
+    DigestWriter digestWriter = new DigestWriter(directories, repositoryName, rule,
+        Preconditions.checkNotNull(ruleSpecificData));
+
+    // Local repositories are fetched regardless of the marker file because the operation is
+    // generally fast and they do not depend on non-local data, so it does not make much sense to
+    // try to cache from across server instances.
+    boolean fetchLocalRepositoryAlways = isFetch.get() && handler.isLocal(rule);
```
Comment (lberki) [inline src/main/java/com/google/devtools/build/lib/rules/repository/RepositoryDelegatorFunction.java]: nit: how about calling this "checkFileSystem"? I spent quite a bit of time trying to figure out why local repositories now check for file system up-to-dateness before realizing that local repositories don't actually get to line 165

Comment (lberki) [inline src/main/java/com/google/devtools/build/lib/rules/repository/RepositoryDelegatorFunction.java]: suggestion: instead of calling `isFetch()` in both of the if statements (lint 164 and line 182) wrap them both in an `if (isFetch()) {` block and potentially extract its contents into a method called e.g. `maybeFetchRepository()`. I found the logic a little hard to follow here, especially due to all the negations (both with the `!` operator and negative variable names and maybe that would help.

File: src/main/java/com/google/devtools/build/lib/rules/repository/RepositoryDelegatorFunction.java [hunk 5]
```diff
@@ -277,38 +275,111 @@ private static Rule getRepository(
...
+    byte[] isFileSystemUpToDate(RepositoryFunction handler, Environment env)
+        throws RepositoryFunctionException, InterruptedException {
       if (!markerPath.exists()) {
         return null;
       }
 
-      String content = FileSystemUtils.readContent(markerPath, StandardCharsets.UTF_8);
+      Map<String, String> markerData = new TreeMap<>();
+      String content;
+      try {
```
Comment (lberki) [inline src/main/java/com/google/devtools/build/lib/rules/repository/RepositoryDelegatorFunction.java]: nit: why not wrap the whole thing into one big try-catch block instead of having three of those?

File: src/main/java/com/google/devtools/build/lib/rules/repository/RepositoryDelegatorFunction.java (hunk 1 above)
Comment (irengrig) [reply src/main/java/com/google/devtools/build/lib/rules/repository/RepositoryDelegatorFunction.java]: Yes, of course you are right, but this is exactly the point of having Preconditions.checkNotNull() at all.
Personally, I also prefer to make a check like "env.valuesMissing()" vs checking some reference for null, so in this case the check "env.valuesMissing()" works in pair with "Preconditions.checkNotNull()" quite well in my opinion in term of expressiveness of code.
= You are saying "but this moment, this value is expected to be computed"

File: src/main/java/com/google/devtools/build/lib/rules/repository/RepositoryDelegatorFunction.java (hunk 4 above)
Comment (irengrig) [reply src/main/java/com/google/devtools/build/lib/rules/repository/RepositoryDelegatorFunction.java]: In general, the computed value **is** equivalent "if it is a local repository, always update".
The fact that later in code for "other" repositories we additionally check if they are up-to-date is not related to the decision about category "local repositories" to be always re-fetched on each execution.

File: src/main/java/com/google/devtools/build/lib/rules/repository/RepositoryDelegatorFunction.java (hunk 2 above)
Comment (irengrig) [reply src/main/java/com/google/devtools/build/lib/rules/repository/RepositoryDelegatorFunction.java]: Corrected

File: src/main/java/com/google/devtools/build/lib/rules/repository/RepositoryDelegatorFunction.java (hunk 4 above)
Comment (irengrig) [reply src/main/java/com/google/devtools/build/lib/rules/repository/RepositoryDelegatorFunction.java]: This is not possible, because in case of isFetch is false we are still checking if the repository already exists and we just return the existing value.

File: src/main/java/com/google/devtools/build/lib/rules/repository/RepositoryDelegatorFunction.java (hunk 5 above)
Comment (irengrig) [reply src/main/java/com/google/devtools/build/lib/rules/repository/RepositoryDelegatorFunction.java]: Done

File: src/main/java/com/google/devtools/build/lib/rules/repository/RepositoryDelegatorFunction.java (hunk 3 above)
Comment (irengrig) [reply src/main/java/com/google/devtools/build/lib/rules/repository/RepositoryDelegatorFunction.java]: Moved the comment into the Javadoc of the method; also, environment has been already passed in the previous variant.
I think that's enough changes already in this refactoring; also, env is mostly used by the handler.
I would rather rename the method isFileSystemUpToDate to something isMarkerFileDirty, WDYT?

Comment (lberki) [reply src/main/java/com/google/devtools/build/lib/rules/repository/RepositoryDelegatorFunction.java]: I thought `isFileSystemUpToDate` is the best possible name: what it does is not checking for the dirtiness of the marker file but whether the FS is up to date by comparing it against the contents of the marker file, if I understand correctly. Maybe `areFileSystemAndMarkerFileConsistent()`?

Comment (irengrig) [reply src/main/java/com/google/devtools/build/lib/rules/repository/RepositoryDelegatorFunction.java]: Renamed to areRepositoryAndMarkerFileConsistent.
I think we should better stick to "repository" concept rather than file system, to keep the focus.

File: src/main/java/com/google/devtools/build/lib/rules/repository/RepositoryDelegatorFunction.java (hunk 4 above)
Comment (lberki) [reply src/main/java/com/google/devtools/build/lib/rules/repository/RepositoryDelegatorFunction.java]: Nevermind. I got confused about the De Morgan identities for a bit.

Comment (lberki) [reply src/main/java/com/google/devtools/build/lib/rules/repository/RepositoryDelegatorFunction.java]: My eyes are *still* bleeding about the number of negations here (`!fetch` and three negations in `!doNotFetchUnconditionally`). Not that I have much better idea.  Let's see. The condition below is:

`repoRoot.exists() && doNotFetchUnconditionally && !(isFetch.get() && isLocal)`

Applying a De Morgan identity we get:

`repoRoot.exists() && doNotFetchUnconditionally && (!isFetch.get() || !isLocal)`

so we do the up-to-dateness check if:

1. the root exists
2. unconditional fetching is not enabled
3. the repository is remote *or* fetching is disabled

Why do we skip the check is fetching is disabled? In that case, do we do anything at all?

Comment (irengrig) [reply src/main/java/com/google/devtools/build/lib/rules/repository/RepositoryDelegatorFunction.java]: If the fetching is disabled, we still check if the repository is up-to-date and then we do nothing.

Comment (irengrig) [reply src/main/java/com/google/devtools/build/lib/rules/repository/RepositoryDelegatorFunction.java]: Sorry, I think two-step logic is better here:
1) we check the category of the repository: local/other.
2) for other repositories, we check if they are already up-to-date.

I added an explaining comment.

Comment (lberki) [reply src/main/java/com/google/devtools/build/lib/rules/repository/RepositoryDelegatorFunction.java]: all the power to you!
